build/parsePrep.c

```diff
@@ -292,18 +292,22 @@ static int doSetupMacro(rpmSpec spec, const char *line)
     }
 
     if (dirName) {
-	spec->buildSubdir = xstrdup(dirName);
+	rpmPushMacro(spec->macros, "buildsubdir", NULL, dirName, RMIL_SPEC);
     } else {
-	rasprintf(&spec->buildSubdir, "%s-%s", 
-		  headerGetString(spec->packages->header, RPMTAG_NAME),
-		  headerGetString(spec->packages->header, RPMTAG_VERSION));
+	char * buildSubdir = rpmExpand("%{?buildsubdir}", NULL);
+	if (buildSubdir[0] == '\0') {
+	    rasprintf(&buildSubdir, "%s-%s",
+		      headerGetString(spec->packages->header, RPMTAG_NAME),
+		      headerGetString(spec->packages->header, RPMTAG_VERSION));
+	    rpmPushMacro(spec->macros, "buildsubdir", NULL, buildSubdir, RMIL_SPEC);
+	}
+	free(buildSubdir);
     }
     /* Mer addition - support --build-in-place */
     if (rpmExpandNumeric("%{_build_in_place}")) {
 	buildInPlace = 1;
-	spec->buildSubdir = NULL;
+	rpmPushMacro(spec->macros, "buildsubdir", NULL, "", RMIL_SPEC);
```

Comment:
This would have to be rpmPopMacro(), as all those %{?buildsubdir} conditionals only test whether the macro is defined at all, not whether it's empty.